src/core/packages/chrome/browser-internal/src/ui/project/sidenav_v2/navigation/navigation.tsx

```diff
@@ -0,0 +1,405 @@
+  );
+  const state = useObservable(state$);
+
+  const memoizedState = useMemo(() => {
+    if (!state) return null;
+    const [navigationTree, navLinks, activeNodes] = state;
+    return toNavigationItems(navigationTree, navLinks, activeNodes);
+  }, [state]);
+
+  return memoizedState;
+};
+
+interface LogoItem {
+  logoType: string;
+  label: string;
+}
+
+/**
+ * Converts the navigation tree definition and nav links into a format for new navigation.
+ *
+ * @remarks
+ *
+ * Structural Assumptions and Mapping
+ *
+ * - Root node (1st level) is used for the "logo" item and application branding
+ * - 2nd level nodes are transformed into primary navigation items:
+ *   - Accordion nodes are flattened (not supported) - their children become primary items
+ *   - Nodes without links that aren't panel openers are treated as section dividers and not supported in new nav - their children are flattened
+ *   - panelOpener nodes create flyout secondary navigation panels, they can't have links directly, but can have sections with links
+ * - 3rd level is used for secondary navigation (children of panelOpener):
+ *   - If all 3rd level items have links, they're treated as menu items and wrapped in a single section
+ *   - If some don't have links, they're treated as section headers with their children becoming menu items
+ * - Footer is limited to 5 items maximum (extras are dropped with warning)
+ *
+ * @param navigationTree
+ * @param navLinks
+ * @param activeNodes
+ */
+export const toNavigationItems = (
+  navigationTree: NavigationTreeDefinitionUI,
+  navLinks: Readonly<ChromeNavLink[]>,
+  activeNodes: ChromeProjectNavigationNode[][]
+): {
+  logoItem: LogoItem;
+  navItems: NavigationStructure;
+} => {
+  // HACK: extract the logo, primary and footer nodes from the navigation tree
+  let logoNode: ChromeProjectNavigationNode | null = null;
+  let primaryNodes: ChromeProjectNavigationNode[] = [];
+  let footerNodes: ChromeProjectNavigationNode[] = [];
+
+  if (navigationTree.body.length === 1) {
+    const firstNode = navigationTree.body[0];
+    if (!isRecentlyAccessedDefinition(firstNode)) {
+      logoNode = firstNode;
+      primaryNodes = firstNode.children ?? [];
+    }
+  } else {
+    warnOnce(
+      `Navigation tree has multiple root nodes. First level should have a single node for the logo and solution name.`
+    );
+  }
+  if (navigationTree.footer?.length === 1) {
+    const firstNode = navigationTree.footer[0];
+    if (!isRecentlyAccessedDefinition(firstNode)) {
+      footerNodes = firstNode.children ?? [];
+    }
+  } else {
+    warnOnce(
+      `Navigation tree footer has multiple root nodes. Footer should have a single node for the footer links.`
+    );
+  }
+
+  const logoItem: LogoItem = {
+    logoType: warnIfMissing(logoNode, 'icon', 'logoKibana') as string,
+    label: warnIfMissing(logoNode, 'title', 'Kibana'),
+  };
+
+  const toMenuItem = (navNode: ChromeProjectNavigationNode): MenuItem[] | MenuItem | null => {
```

Comment:
`toMenuItem` can be a top-level function; it does not need to be enclosed

Reply:
Also, a preference. `toMenuItem` is just an internal implementation detail of that mapper, and we don't really need to keep it outside.